src/transform/src/canonicalization/mod.rs

```diff
@@ -21,3 +21,56 @@ mod topk_elision;
 pub use flatmap_to_map::FlatMapToMap;
 pub use projection_extraction::ProjectionExtraction;
 pub use topk_elision::TopKElision;
+
+use crate::TransformCtx;
+use crate::TransformError;
+use mz_expr::MirRelationExpr;
+
+/// Orders the keys in a `Reduce`.
+///
+/// This ordering is best thought of as a "logical" transformation, in that
+/// it aims to canonicalize the representation without (yet) worrying about
+/// physical properties like the order of arrangement keys (which can help
+/// to avoid re-arranging data).
+#[derive(Debug)]
+pub struct ReduceKeyOrder;
+
+impl crate::Transform for ReduceKeyOrder {
+    #[tracing::instrument(
+        target = "optimizer",
+        level = "debug",
+        skip_all,
+        fields(path.segment = "reduce_key_order")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: &mut TransformCtx,
+    ) -> Result<(), TransformError> {
+        // Visit looking for `Reduce` expressions with keys not in canonical order.
+        relation.visit_pre_mut(|expr| {
+            if let MirRelationExpr::Reduce {
+                group_key: keys,
+                aggregates,
+                ..
+            } = expr
+            {
+                let old_keys = keys.clone();
+                keys.sort();
```

Comment:
How about also deduplicating them while we are at it? (And then it could be called `ReduceKeyCanonicalization` instead of `ReduceKeyOrder`.)

Reply:
Well, I did this (and ordered aggregates) and it's now +/-1.2KLOC. Sorry. =/